### PR TITLE
feat(3d-tiles): 3tz loader

### PR DESCRIPTION
**Pull request by dariaterekhovaae**
Ready to merge. TZ3 loader fully implemented

### CODE REVIEW

File: modules/loader-utils/test/zip-utils/data-view-provider.spec.ts
Comment (ibgreen) [inline modules/loader-utils/test/zip-utils/data-view-provider.spec.ts:2]: We normally don't include .js extension in imports.
```suggestion
import {DATA_ARRAY} from '@loaders.gl/i3s/test/data/test.zip';
```

File: modules/loader-utils/test/zip-utils/data-view-provider.spec.ts
Comment (ibgreen) [inline modules/loader-utils/test/zip-utils/data-view-provider.spec.ts:3]: Not for this PR: Should this class (and colleagues) just be called `DataViewFile`? 
- The "provider" patter in JavaScript React is a kind of component that let's components communicate without passing props to each other. 
- Ignoring the fairly established provider pattern perhaps seems to me a provider could be some kind of factory that provides / lets you create objects.
- But probably not the objects (files) themselves?

File: modules/loader-utils/test/zip-utils/end-of-central-directory.spec.ts
Comment (ibgreen) [inline modules/loader-utils/test/zip-utils/end-of-central-directory.spec.ts:2]: Again drop the .js extension if possible

File: modules/loader-utils/test/zip-utils/local-file-header.spec.js
Comment (ibgreen) [inline modules/loader-utils/test/zip-utils/local-file-header.spec.js:2]: ditto

File: modules/loader-utils/test/zip-utils/search-from-the-end.spec.js
Comment (ibgreen) [inline modules/loader-utils/test/zip-utils/search-from-the-end.spec.js:2]: ditto

**Review by ibgreen — COMMENTED** [review]
So this PR doesn't really implement MAXAR_content_3tz but just prepares for it?

My main concern is that the loader-utils module is included by almost all loaders so we want to be careful that we don't add big functionality that is only used by one or few loaders.

I agree that the so called "FileProviders" can fit into loaders.gl, but I think zip file support is a step too far.

Could we put the zip file support in the loaders.gl zip module instead?

**Comment by ibgreen** [comment]
If you are waiting for approval, please move this PR out of draft status and notify me.

File: modules/3d-tiles/src/index.ts
Comment (ibgreen) [inline modules/3d-tiles/src/index.ts:5]: Nit: Split type and symbol exports. 
```suggestion
export type {TZ3LoaderOptions} from './tz3-loader';
export {TZ3Loader} from './tz3-loader';
```
Thanks for adding loader types here. It would be nice to go back to other 3d tiles loaders and add options types and export them here.

File: modules/3d-tiles/src/tz3/tz3-archive.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3/tz3-archive.ts:78]:
```suggestion
      const compression = COMPRESSION_METHODS[localFileHeader.compressionMethod.toString()];
      if (!compression) { 
```

File: modules/compression/src/lib/deflate-compression.ts
Comment (ibgreen) [inline modules/compression/src/lib/deflate-compression.ts:11]:
```suggestion
  /** Please explain this option in a comment here */
  raw?: boolean;
```

File: modules/i3s/src/lib/parsers/parse-slpk/parse-slpk.ts
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-slpk/parse-slpk.ts:1]: Nit: These changes seem unrelated? Could be a separate PR?

**Review by ibgreen — COMMENTED** [review]
There is just too much going on in this PR. I can't review everything. If possible please split into a couple of unrelated PRs.

Finally I would want to see some user documentation updates for the new 3tz format and loader, also don't forget to update `docs/whats-new.md`

**Comment by dariaterekhovaae** [comment]
> There is just too much going on in this PR. I can't review everything. If possible please split into a couple of unrelated PRs.

I'll create some new pr's instead of this one

File: modules/3d-tiles/src/tz3-loader.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3-loader.ts:20]: Nit: Are we better off calling this `Tiles3DArchiveLoader`? 

I suspect it will make this loader more easily understood by users who have never heard about `3tz` and even more confusingly, we have to reorder the letters to `tz3`...

3tz feels more like the file extension. https://github.com/erikdahlstrom/3tz-specification

File: modules/3d-tiles/src/tz3-loader.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3-loader.ts:25]: I wonder if e.g. Cesium has some vendor specific mime type for this? Worth Googling a bit?

File: modules/3d-tiles/src/tz3-loader.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3-loader.ts:38]: - Type out the return type if possible
- Too much going on one line for my tastes. I'd break this into two lines for readability:
```suggestion
  const archive = await parse3tzFromProvider(new DataViewFile(new DataView(data))));
  archive.getFile(...);
```

File: modules/3d-tiles/src/tz3-loader.ts
Comment (ibgreen) [reply modules/3d-tiles/src/tz3-loader.ts:20]: Also since we only load a single file, maybe call it `Tiles3DArchiveFileLoader`. Had a similar discussion with @belom88 around the Zip loader.

File: modules/3d-tiles/src/tz3/tz3-archive.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3/tz3-archive.ts:20]: Does something like this work? Compression module does provide a `NoCompression` class with the same interface.
```suggestion
const COMPRESSION_METHODS: {[key: number]: CompressionHandler} = {
  /** No compression */
  0: new NoCompression().decompress,
  /** Deflation */
  8: new DeflateCompression({raw: true}).decompress
};
```

File: modules/3d-tiles/src/tz3/tz3-archive.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3/tz3-archive.ts:25]:
```suggestion
export class Tiles3DArchive {
```

File: modules/3d-tiles/src/tz3/tz3-archive.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3/tz3-archive.ts:38]: If at all possible, we do not want to expose Node.js buffer's in the loaders.gl API. They can technically be used in the browser but only because the bundler pulls in a big polyfill into the built code bundle, plus the applications now become more complex as they have to deal with both ArrayBuffer/Typed Arrays and Buffer objects for binary data.

Prefer ArrayBuffer if possible.

File: modules/3d-tiles/src/tz3/tz3-parser.ts
Comment (ibgreen) [inline modules/3d-tiles/src/tz3/tz3-parser.ts:21]: Callback parameters is not something we have been using in loaders.gl. 
Perhaps passing in a `log` instance could solve the problem here?

**Review by ibgreen — APPROVED** [review]

**Comment by dariaterekhovaae** [comment]
I'll fix all of the comments in a separate pr

**Pull request merged**
